### PR TITLE
scalar_in_linter() for x %in% 1

### PR DESCRIPTION
**Pull request by MichaelChirico**
Part of #884

### CODE REVIEW

**Comment by AshesITR** [comment]
In the presence of NAs, `x %in% "a"` is much better than `is.na(x) | x == "a"` IMO...

**Comment by AshesITR** [comment]
We could lint `x %in% NA` though to suggest `is.na(x)`.

**Comment by MichaelChirico** [comment]
> In the presence of NAs, x %in% "a" is much better than !is.na(x) & x == "a" IMO...

I prefer the latter for making missing behavior explicit... if you prefer the former it's best not to use this linter I'd say. We can tweak the documentation to reflect that it may be up to taste, WDYT?

**Comment by MichaelChirico** [comment]
> We could lint `x %in% NA` though to suggest `is.na(x)`.

I'd make that a message customization as a follow-up

**Comment by AshesITR** [comment]
> > In the presence of NAs, x %in% "a" is much better than !is.na(x) & x == "a" IMO...
> 
> 
> 
> I prefer the latter for making missing behavior explicit... if you prefer the former it's best not to use this linter I'd say. We can tweak the documentation to reflect that it may be up to taste, WDYT?

Fine by me, though I might turn on the RHS = NA case specifically.

**Comment by AshesITR** [comment]
> > We could lint `x %in% NA` though to suggest `is.na(x)`.
> 
> 
> 
> I'd make that a message customization as a follow-up

Keep in mind that will generate incorrect advice since the behavior of == differs from %in% there.
It would be nice to restrict this linter to only this case btw because `x %in% NA` vs. `is.na(x)` is more clearly a lint IMO.
The logic should be quite simple - extract the constant as an R value and test it using `is.na()` or just hard-code the possible NAs, maybe offering capabilities for extension to NAs from other packages such as `lubridate::NA_Date_`.

File: R/scalar_in_linter.R [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+#'
+#' @evalRd rd_tags("scalar_in_linter")
+#' @seealso [linters] for a complete list of linters available in lintr.
+#' @export
+scalar_in_linter <- function() {
+  # TODO(#2085): Extend to include other cases where the RHS is clearly a scalar
+  # NB: all of logical, integer, double, hex, complex are parsed as NUM_CONST
+  xpath <- "
+  //SPECIAL[text() = '%in%' or text() = '%chin%']
+    /following-sibling::expr[NUM_CONST or STR_CONST]
```
Comment (AshesITR) [inline R/scalar_in_linter.R:18]: We should either exclude NAs from this XPath or implement the branching right away I think. What do you think about it?

Comment (MichaelChirico) [reply R/scalar_in_linter.R:18]: Oh, I missed your point earlier I think -- the recommendation to use `==` is just wrong for NA on the RHS (it's an `equals_na_linter()` violation too). I think the reason this escaped us so far is we'd already fixed the handful of `equals_na_linter()` hits in our codebase by the time this was written.

Agree it should be addressed up-front.

Comment (MichaelChirico) [reply R/scalar_in_linter.R:18]: Done. For now, we customize the message for `NA` on the RHS. More parameterization to follow in subsequent enhancements.

File: inst/lintr/linters.csv [hunk 2]
```diff
@@ -71,6 +71,7 @@ redundant_equals_linter,best_practices readability efficiency common_mistakes
 redundant_ifelse_linter,best_practices efficiency consistency configurable
 regex_subset_linter,best_practices efficiency
 routine_registration_linter,best_practices efficiency robustness
+scalar_in_linter,readability consistency best_practices efficiency
```
[the remark this answers is not in the document]
Comment (MichaelChirico) [reply inst/lintr/linters.csv:74]: `==` faster, especially if we know NA's not involved

```r
library(microbenchmark)

## CHARACTER
x = sample(letters, 1e7, TRUE)
microbenchmark(times = 10, x == "a", x %in% "a")
# Unit: milliseconds
#        expr       min        lq      mean   median        uq       max neval cld
#    x == "a"  40.56016  44.49466  58.97258  48.5303  72.51743  88.75041    10  a 
#  x %in% "a" 149.08170 169.48051 220.53054 197.9211 247.93525 417.98458    10   b
microbenchmark(times = 10, !is.na(x) & x == "a", x %in% "a")
# Unit: milliseconds
#                  expr      min        lq     mean    median       uq      max neval cld
#  !is.na(x) & x == "a"  73.3217  86.99122 142.9826  91.88972 224.6992 305.4078    10  a 
#            x %in% "a" 150.9287 176.25386 260.3127 250.10186 363.0894 387.9572    10   b

## INTEGER
x = sample(100, 1e7, TRUE)
microbenchmark(times = 10, x == 12, x %in% 12)
# Unit: milliseconds
#       expr      min        lq      mean    median        uq       max neval cld
#    x == 12  26.6641  28.44449  31.35452  30.51409  33.78526  37.81461    10  a 
#  x %in% 12 218.5079 227.94896 230.85022 231.09179 231.85876 248.32053    10   b
microbenchmark(times = 10, !is.na(x) & x == 12, x %in% 12)
# Unit: milliseconds
#                 expr       min        lq     mean   median       uq      max neval cld
#  !is.na(x) & x == 12  67.37458  76.47258 125.0347 102.4130 131.5533 280.6254    10  a 
#            x %in% 12 229.43924 252.39638 306.9022 275.8865 385.9531 417.3146    10   b

## NA RHS CASE
is.na(x) <- sample(1e7, 100)
microbenchmark(times = 10, is.na(x), x %in% NA_integer_)
# Unit: milliseconds
#                expr        min         lq      mean     median        uq      max neval cld
#            is.na(x)   6.843287   7.100379  22.87707   8.405525  14.53488 140.7589    10  a 
#  x %in% NA_integer_ 103.639572 104.065773 109.26014 109.232751 112.42047 117.8867    10   b
```

NB: also tried throwing in some `NA`s to those examples without much effect (`is.na(x) <- sample(1e7, 100)`).

[the remark this answers is not in the document]
Comment (MichaelChirico) [reply inst/lintr/linters.csv:74]: > Should we add common_mistakes if we lint NAs?

Not a bad idea, but do you think it's enough to cover that mistake with `equals_na_linter()` (which is already a default):

https://github.com/r-lib/lintr/blob/430123122ca91c4642d21d32ca9a993918a8c1cf/R/zzz.R#L21

[the remark this answers is not in the document]
Comment (MichaelChirico) [reply inst/lintr/linters.csv:74]: Maybe we should lint `x %in% NA` under `equals_na_linter()`, come to think of it... WDYT?

[the remark this answers is not in the document]
Comment (AshesITR) [reply inst/lintr/linters.csv:74]: That sounds like a better place to put it. So we should exclude it here.

[the remark this answers is not in the document]
Comment (MichaelChirico) [reply inst/lintr/linters.csv:74]: Nothing stopping `%in%` from needing dispatch as well, e.g. any session that even transitively loads {raster}:

https://github.com/rspatial/raster/blob/694aeb5bdff13162cc7fae5fbcb82673c6a379f7/R/AAgeneric_functions.R#L26

[the remark this answers is not in the document]
Comment (MichaelChirico) [reply inst/lintr/linters.csv:74]: > That sounds like a better place to put it.

Filed under #2088.

[the remark this answers is not in the document]
Comment (MichaelChirico) [reply inst/lintr/linters.csv:74]: OK, currently `x %in% NA` will not lint from `scalar_in_linter()`, and will later lint from `equals_na_linter()`. PTAL.

File: tests/testthat/test-scalar_in_linter.R [hunk 3]
```diff
@@ -0,0 +1,32 @@
+test_that("scalar_in_linter skips allowed usages", {
+  linter <- scalar_in_linter()
+
+  expect_lint("x %in% y", NULL, linter)
```
Comment (AshesITR) [inline tests/testthat/test-scalar_in_linter.R:4]: Can we add a test for NA?

Comment (AshesITR) [reply tests/testthat/test-scalar_in_linter.R:4]: Related: what's the parse data specifically for `NA_character_`?

Comment (MichaelChirico) [reply tests/testthat/test-scalar_in_linter.R:4]: `<NUM_CONST>NA_character_</NUM_CONST>`, oddly enough, which is straight from the parser:

```r
getParseData(parse(text = "NA_character_"))
  line1 col1 line2 col2 id parent     token terminal          text
1     1    1     1   13  1      2 NUM_CONST     TRUE NA_character_
2     1    1     1   13  2      0      expr    FALSE              
```

Comment (MichaelChirico) [reply tests/testthat/test-scalar_in_linter.R:4]: Test added.

File: tests/testthat/test-scalar_in_linter.R [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+  expect_lint("y %in% c('a', 'b')", NULL, linter)
+  expect_lint("c('a', 'b') %chin% x", NULL, linter)
+  expect_lint("z %in% 1:3", NULL, linter)
+  # scalars on LHS are fine (often used as `"col" %in% names(DF)`)
+  expect_lint("3L %in% x", NULL, linter)
+
+  # this should be NA, but it more directly uses the "always TRUE/FALSE, _not_ NA"
```
[the remark this answers is not in the document]
Comment (MichaelChirico) [reply tests/testthat/test-scalar_in_linter.R]: Yes 👍

**Review by AshesITR — APPROVED** [review]

**Pull request merged**